Atomic.update (was: Atomic.modify, Atomic.modify_get)

gasche opened the pull request:
```ocaml
val modify : ('a -> 'a) -> 'a t -> unit
val modify_get : ('a -> 'b * 'a) -> 'a t -> 'b
```

Many uses of `Atomic.compare_and_set` are in retry loops that can be factored out by these new functions. Consider:

https://github.com/ocaml-multicore/ocaml-multicore/blob/4469f89/stdlib/printexc.ml#L268-L272
```ocaml
(* old version *)
let rec register_printer fn =
  let old_printers = Atomic.get printers in
  let new_printers = fn :: old_printers in
  let success = Atomic.compare_and_set printers old_printers new_printers in
  if not success then register_printer fn

(* new version *)
let register_printer fn = Atomic.modify (List.cons fn) printers
```

https://github.com/ocaml-multicore/ocaml-multicore/blob/4469f89/testsuite/tests/parallel/mctest.ml#L222-L226
```ocaml
(* old version *)
  let rec finish () =
    let v = Atomic.get counter in
    if not (Atomic.compare_and_set counter v (v+1)) then finish ();
    if v + 1 = procs then exit 0

(* new version *)
  let finish () =
    (* this could also be done with the more specialized Atomic.fetch_and_add *)
    let next = Atomic.modify_get (fun v -> v+1, v+1) in
    if next = procs then exit 0
```

https://github.com/ocaml-bench/sandmark/blob/642d16edeb1ad68e5fbe5019035f6fc4f8223dea/benchmarks/multicore-structures/treiber_stack.ml
```ocaml
(* old version *)
let rec add_node s n =
  let old_top = Atomic.get s.top in
  Atomic.set n.next old_top ;
  if Atomic.compare_and_set s.top old_top (Some n) then () else add_node s n

(* new version *)
let add_node s n =
  Atomic.modify (fun old_top ->
    Atomic.set n.next old_top ; Some n
  ) s.top
```

https://github.com/dune-universe/dune-universe/blob/eef415c39aa4533cf5fad448e2b9a31e15ee03b8/packages/uring.0.1/lib/uring/uring.ml
```ocaml
(* old *)
let rec update_gc_roots fn =
  let old_set = Atomic.get gc_roots in
  let new_set = fn old_set in
  if not (Atomic.compare_and_set gc_roots old_set new_set) then
    update_gc_roots fn

(* new *)
let update_gc_roots fn = Atomic.modify fn gc_roots
```

Remark:
- Some more elaborate uses of `compare_and_set` are in more complex retry loops (see for example the [Domainslib.Chan](https://github.com/ocaml-multicore/domainslib/blob/master/lib/chan.ml) implementation)
- The current definitions do not call `cpu_relax ()` or any other backoff logic. Some uses in the wild would be expressible if we parametrized over the backoff function. I don't know if this is a choice that users should be forced to make.

gasche committed:
Atomic.modify, Atomic.modify_get

stedolan commented:
This looks useful to me, but I suggest a change to the documentation. The current text:
```
    [modify f r] computes a new value for [r] by applying [f] to its
    current value, sets this new value or retries (calling [f] again)
    if [r] was concurrently changed to a physically different value.
```
sounds like it includes a solution to the [ABA problem](https://en.wikipedia.org/wiki/ABA_problem). The text should be changed to make it clear it does not. (Maybe "if [r] now holds a physically different value" rather than "if [r] was concurrently changed to a physically different value"? Or maybe an explicit disclaimer. Or, alternatively, implement a `modify` that does solve ABA, although the types will have to be slightly different)

xavierleroy commented:
> implement a modify that does solve ABA

This sounds wonderful!  (I hate ABA problems with a passion...)  How would you go about that?

xavierleroy commented:
> The current definitions do not call cpu_relax () or any other backoff logic. Some uses in the wild would be expressible if we parametrized over the backoff function. I don't know if this is a choice that users should be forced to make.

What about an optional argument?  Then we can argue about the default backoff function...

gasche commented:
> What about an optional argument? Then we can argue about the default backoff function...

Why not indeed. I'm waiting for someone to jump in and claim that it is almost always the wrong default to not backoff, and that we should *force* users to think about this by making it non-optional. But I don't have enough experience to make that claim. (The code I found in the wild often does not backoff, which may be an argument one way or another.)

c-cube commented:
I might be very naive here, but is ABA a problem in OCaml here? Quoting from the [wikipedia page](https://en.wikipedia.org/wiki/ABA_problem):

> In multithreaded computing, the ABA problem occurs during synchronization, when a location is read twice, has the same value for both reads, and "value is the same" is used to indicate "nothing has changed".
> […]
> A common case of the ABA problem is encountered when implementing a lock-free data structure. If an item is removed from the list, deleted, and then a new item is allocated and added to the list, it is common for the allocated object to be at the same location as the deleted object due to MRU memory allocation. A pointer to the new item is thus often equal to a pointer to the old item, causing an ABA problem. 

However, OCaml values are a bit special (assuming no mutation, which one wouldn't mix with atomics anyway). Since `modify` keeps a handle on the old value, it cannot be GC'd, so any new value will be physically distinct. Hence either the value is an unboxed value, in which case physical equality is enough; or it's heap allocated and then you never go back to "A" unless it's the old object. For example, with the stack, you can't get back the same cons cell with a different content in another thread, since the old cons cell is still alive.

Am I missing something?

stedolan commented:
> This sounds wonderful! (I hate ABA problems with a passion...) How would you go about that?

When you have a GC lying around, you can ensure freshness by making a new allocation each time. Below I'm going to use a "mutable" field, even though I won't mutate it, to prevent the optimiser getting any ideas. (Or, more formally, because physical equality is only determinstically defined for allocations containing mutable fields):

```ocaml
type 'a box = { mutable thing : 'a }
type 'a aba_atomic = 'a box Atomic.t

let read (a : 'a aba_atomic) =
  (Atomic.get a).thing

let rec modify f (a : 'a aba_atomic) =
  let v_old = Atomic.get r in
  let v_new = { thing = f v_old.thing } in
  if Atomic.compare_and_set r v_old v_new
  then ()
  else modify f r
```

If ABA occurs, then the contents of the ref will not be physically equal to `v_old`, because there's a fresh allocation on each `modify`.

@c-cube you're right, as long as you're consing something on each `modify`. For that reason, you don't need the pattern above for such cases, although the pattern above works in all cases by forcing a cons each time.

c-cube commented:
Thank you @stedolan, that makes sense to me indeed. I think this particular pattern of `Atomic.modify` should be labelled clearly as something that works well with immutable values. If you have internal mutation, a mutex is probably best.

People have long claimed that one of the benefits of FP is that it's easier to parallelize, and this might be a concrete case where this is true! :grin:

gasche commented:
Fun story time: today I was writing some Atomic code, and I wanted to atomically cons a value to a list, so I thought about submitting an `Atomic.cons` PR. Then I realized that `Atomic.modify` was more expressive, and:

```console
$ git checkout -b Atomic.modify
fatal: a branch named 'Atomic.modify' already exists
```

[🤦](https://emojipedia.org/person-facepalming/)

My intuition is that at least `Atomic.cons` would be very nice to have *even in 5.0*, and `Atomic.modify` is still a nice generalization. Last time we got lost in a discussion of the ABA problem, but could we maybe try again and see if we can come to a decision this time?

Re. ABA: I propose to just reformulate the documentation to make it clear that we don't solve it.
Re. cpu_relax: I propose to do without it for now, and wait to see what happens. (I dislike adding optional arguments after the fact, but I also dislike adding optional arguments in the first place, oh well...)

xavierleroy commented:
It's too late to add features to 5.0.  This PR is going to compete with the many that could go into 5.1 if only we had more reviewing power.

bikallem commented:
This looks like a good addition to the stdlib `Atomic` module. I copied the `modify` function defined here and it seems to work quite well. Overall useful set of functions to have.

gadmm commented:
I agree with @c-cube here regarding the ABA problem. @stedolan's example seems to prevent the case `v_old == v_new` but this is not a concern here (if `f ≝ fun x -> x` then the original works as expected). The issue where the `v_old` as a physical location is modified to denote a different value (typically by doing `free` followed by `malloc` in non-GC-ed languages) is avoided here.

There can be more higher-level incarnations of the ABA problem (involving mutable OCaml values), but as far as the ABA problem of CAS on `Atomic.t` is concerned, I believe that this implementation or any other naive one avoids it.

gadmm reviewed (commented):
Nobody's chiming in but I think we're nearly there. 
1. As mentioned, the ABA problem is not an issue and there is no need to do something documentation-wise.
2. The backoff question is also a non-issue. This kind of CAS loop is typically based on the reasoning that failure means that another thread has successfully progressed, so you want to retry ASAP. In case of contention, the strategies to improve scaling typically do not involve executing the `pause` instruction (but instead, for instance, distributing the load among several atomic variables). This is completely different from when you implement a spinlock.
3. Another issue that I see is that there might be an overhead of the function call inside the critical section. For short `f` this overhead might be substantial (`modify` would typically be called for short `f`s). Have you looked at generated code to see if the critical sections were worse than hand-rolled code? If it is, is there something that could be done wrt. inlining to bring such guarantee? If one cannot guarantee that the critical section is as efficient as possible, using `modify` could become an anti-pattern.
  If inlining does not work, one could imagine letting `f` be responsible for calling `Atomic.get` with signature `val modify : ('a t -> 'a) -> 'a t -> unit` (thus the load happens after the function call; there's still the overhead of returning). As an added benefit, `f` would gain the choice between calling `Atomic.get` and future relaxed variants of it.

This is basically an "accept" but I think that the point 3. requires further discussion.

Please rebase: update `@since` (edit: apologies for forgetting the backticks and pinging a github user) and adapt to removal of `CamlinternalAtomic` (you might need to abandon the idea to use `modify` in modules loaded before `Atomic`).

polytypic commented:
In recent [updates I made to the kcas library](https://github.com/ocaml-multicore/kcas/pull/22), I opted for a slightly different approach compared to the `modify_get` proposed here. I called it [`update`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Loc/index.html#val-update) and it has the following signature:

```ocaml
val update : ?backoff:Backoff.t -> 'a t -> ('a -> 'a) -> 'a
```

The value returned by `update` is the value that was previously in the location.

You could interpret the `'a t` as either an `'a Atomic.t` or as a `'a Kcas.Loc.t`.

Let me explain why I ended up with this signature.

First of all, previous versions of kcas had two operations [`try_map`](https://ocaml-multicore.github.io/kcas/0.1.8/kcas/Kcas/index.html#val-map) and [`map`](https://ocaml-multicore.github.io/kcas/0.1.8/kcas/Kcas/index.html#val-try_map) that `update` now replaces:

```diff
-match Kcas.map r (fun c -> if condition c then None else Some (next c)) with
-| Success c -> (* success *)
-| Failed | Aborted -> (* aborted *)
+match Loc.update r (fun c -> if condition c then raise Exit else next c) with
+| c -> (* success *)
+| exception Exit -> (* aborted *)
```

```diff
-match Kcas.try_map r (fun c -> if condition c then None else Some (next c)) with
-| Success c -> (* success *)
-| Failed -> (* failed *)
-| Aborted -> (* aborted *)
+let c = Loc.get r in
+if condition c then (* aborted *)
+else if Loc.compare_and_set r c (next c) then (* success *)
+else (* failure *)
```

In other words, both `map` and `try_map` can be expressed pretty easily and just as concisely using `update`.

Now, some specific notes about `update`. First of all, the core signature is perhaps the simplest possible for this kind of use:

```ocaml
'a t -> ('a -> 'a) -> 'a
```

Second, the signature does not force the function given to `update` to perform an allocation. Allocations are cheap in OCaml, but they are not free.

Third, that `update` returns the previous value, allows one to compute whatever one wants based on that value and do it outside of the spin loop. Consider this example from the PR:

```ocaml
let () =
  let r = Atomic.make 1 in
  let v =
    Atomic.modify_get (fun i ->
      begin
        (* simulate concurrent modifications *)
        if i < 10 then Atomic.incr r;
      end;
      string_of_int i, i + 1
    ) r
  in
  assert (v = "10")
```

The call to `string_of_int` is performed potentially many times and it is performed during the critical period between `Atomic.get` and `Atomic.compare_and_set`. It is better to move that outside of that critical period.

Using an `Atomic.update` that would look like this:

```ocaml
let () =
  let r = Atomic.make 1 in
  let i =
    Atomic.update r (fun i ->
      begin
        (* simulate concurrent modifications *)
        if i < 10 then Atomic.incr r;
      end;
      i + 1
    )
    |> string_of_int
  in
  assert (v = "10")
```

Now the `string_of_int` computation is moved outside of the critical period.

In general, one can replace `modify` using `update` as:

```diff
-Atomic.modify (fun prev -> (result_of prev, next_of prev)) atomic
+Atomic.update atomic next_of |> result_of
```

The new kcas also offers a transactional [`Tx`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Tx/index.html) API, that offers some further shorthands. One of them is [`update_as`](https://ocaml-multicore.github.io/kcas/doc/kcas/Kcas/Tx/index.html#val-update_as) and using an `Atomic.update_as` the above example could be written as:

```ocaml
let () =
  let r = Atomic.make 1 in
  let i =
    Atomic.update_as string_of_int r (fun i ->
      begin
        (* simulate concurrent modifications *)
        if i < 10 then Atomic.incr r;
      end;
      i + 1
    )
  in
  assert (v = "10")
```

Note that with the transaction mechanism it is advantageous to minimize the number of accesses to the transaction log.  That is the main motivation for having `update_as` and several other access operations.  (In fact, `update_as` is the most generic variant and can express all the other available operations for accessing a location inside a transaction.)

polytypic commented:
Backoff was also mentioned.

The update to kcas also includes a new backoff mechanism implementation.  See the [signature](https://github.com/ocaml-multicore/kcas/blob/fc70e5b944a7a7c61980cd247326f21ae7f66cd8/src/backoff.mli) and the [implementation](https://github.com/ocaml-multicore/kcas/blob/fc70e5b944a7a7c61980cd247326f21ae7f66cd8/src/backoff.ml).

The special feature of this new design is that it does not require allocations.  The `Backoff.t` type is just an `int` that is packed to include three 5-bit fields.

So, having an optional `?backoff:Backoff.t` argument is (almost) free.  The idea isn't, however, that backoff would be optional.  The idea is that the caller may optionally specify a backoff with desired configuration.

gasche commented:
Shouldn't your Backoff logic actually `usleep` once the backoff time gets large enough, to also improve fairness on contended systems? This is what the runtime backoff code does, in a somewhat obscure way: 
the first `Max_spins = 1000` iterations are just a cpu_relax, and then we `usleep`.

https://github.com/ocaml/ocaml/blob/1c5e7488430196341cae155e119719d0fa69d800/runtime/caml/platform.h#L86-L93

https://github.com/ocaml/ocaml/blob/1c5e7488430196341cae155e119719d0fa69d800/runtime/platform.c#L223-L240

polytypic commented:
> @gadmm  This kind of CAS loop is typically based on the reasoning that failure means that another thread has successfully progressed, so you want to retry ASAP. 

Hmm... Here is an excellent blog post on [Exponential Backoff and Jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/).  As explained in that blog post, the idea with a randomized exponential backoff is to distribute the retries in time.  Naïve "let's retry now as someone else already made their update" actually leads directly to the situation where everyone retries at the same time.

polytypic commented:
> Shouldn't your Backoff logic actually usleep once the backoff time gets large enough, to also improve fairness on contended systems? 

I think in some context that would be fine.  Usually with spin-loops the hope is that it gets done quickly.  I would probably avoid using this specific kind of backoff mechanism outside the tightest atomic update loops.  For longer waits one also wants to take into account things like user space fibers etc.  In cases where backoff times are substantially longer one would probably just directly go to sleep and/or let the scheduler run some other work.

gadmm commented:
@polytypic This is a blog post about contention with DynamoDB transactions and I am not convinced that it is relevant for typical use-cases of CAS loops over `Atomic.t`.

polytypic commented:
> @polytypic This is a blog post about contention with DynamoDB transactions and I am not convinced that it is relevant for typical use-cases of CAS loops over `Atomic.t`.

Backoff is used for fundamentally the same reasons in both contexts.  See e.g. [this blog post](https://geidav.wordpress.com/tag/exponential-back-off/) or [Algorithms for Scalable Synchronization on Shared-Memory Multiprocessors](https://www.cs.rochester.edu/u/scott/papers/1991_TOCS_synch.pdf):

> The total amount of network traffic caused by busy-waiting on a test-and-set lock can be reduced further by introducing delay on each processor between consecutive probes of the lock. The simplest approach employs a constant delay; more elaborate schemes use some sort of backoff on unsuccessful probes. Anderson [5] reports the best performance with exponential backoff; our experiments confirm this result.

I would perhaps distinguish between two fundamentally different situations:
1. You are implementing a specific algorithm where you know that it just makes sense to retry immediately or to use some other backoff strategy.
2. You are providing a general purpose mechanism and have no knowledge of the context.

In the former kind of case you might e.g. know that the number of "participants" (domains, fibers, whatever) that will be updating an atomic is going to be very small (maybe just 1 or 2 besides you).  So, it is fine to just retry immediately.  In those cases I would recommend writing the loop explicitly.  In some such cases you don't even need a loop, you can use `exchange` and e.g. make sure the `exchange` is to a terminal state.  Or you might have additional information and provide some sort of proportional backoff based on that knowledge:

> In this case, however, exponential backoff is clearly a bad idea. Since processors acquire the lock in FIFO order, overshoot in backoff by the first processor in line will delay all others as well, causing them to back off even farther. Our experiments suggest that a reasonable delay can be determined by using information not available with a test-and-set lock: namely, the number of processors already waiting for the lock.

In the latter case, well, you don't know.  What you do know, however, is that immediate retry has the potential to cause a lot of shared memory traffic.  Randomized exponential backoff is a reasonable way to avoid that.

gadmm commented:
That article and that blog post discuss the implementation of spin locks. You are wrong in thinking that they should be relevant to the present PR. It would better help you making your points if you avoid having us read irrelevant papers.

polytypic commented:
@gadmm You haven't so far provided a single shred of evidence to back your arguments.  Here is another paper, [A Scalable Lock-free Stack Algorithm](https://people.csail.mit.edu/shanir/publications/Lock_Free.pdf), that discusses a lock-free stack algorithms using an elimination array.  What is interesting, however, is that it also benchmarks against a Treiber stack with and without exponential backoff.  The version with exponential backoff performs better than the version without.  At any rate, I have said what I wanted to.  Good luck!

gasche commented:
Your last message does support the idea that low-level concurrent stacks (and probably other CAS-based shared structures) need a backoff strategy for scalability in high-contention scenarios. I think that @gadmm made a technically correct point that busy-wait loops are very different beasts that obviously benefit from contention-avoidance -- in an unpleasant way.
I was also interested in the DynamoDB post on the importance of backoff-time randomization. One thing that was a bit worrying in the DynamoDB is the cost of the backoff strategy on throughput. In the paper you cite on the exponential-backoff stack, there is no particular throughput cost to backoff on their benchmarks, so this is more reassuring. This difference does support the idea that the difference in scale between DynamoDB and typically lockfree data structures means that we cannot fully transfer performance intuitions from one to another.

I was already rather convinced that having backoff by default is a reasonable approach. I wonder if we can find a backoff strategy that is "good enough" to be unconditionally enabled by a standard `Atomic.modify`-style function. (It doesn't need to consistently outperform all approaches, but it should be unsurprising and decent for broad usage.) I would rather not make the API more complex by making the backoff scheme configurable, if we can avoid it.

@polytypic thanks for your pointer on your `update` function. Just like fetch_add vs. add_fetch, we could have a `get_modify` variant of `modify_get` that returns the previous value. (I haven't looked at this in full details yet so I can't comment on the rest of your API choices.)

Unrelatedly: I'm happy to learn that the ABA issues appear to have magically solved themselves since I opened the PR thanks to @c-cube. ABA is hard to think about for a non-expert like me and I'm enthusiastic about not thinking about it at all.

gadmm commented:
This last paper was the first one that showed up for me on duckduckgo for "lock-free data structure exponential backoff". To me it has been an example where it is better to spread accesses in space rather than time (here with an "elimination array"). I still think it makes my point, even if it displays a Treiber stack with exponential backoff as a control. (Besides, the paper does not really try to justify the control with backoff in time. I am worried about fairness but the paper only shows average latency—where should I start if I cannot even read a proper paper that presents the claims of people who defend the backoff in time?)

But I think this is missing the point anyway. The question is not whether in a given highly-contended situation adding a backoff (in time) can reduce cache coherence traffic and improve throughput (by giving priority to the winning thread). It is also not about helping programmers with specialised data structures (who can trivially write their own specialised loop). Most of the time you write basic CAS loops with uncontended situations in mind; otherwise you first try to rephrase your program to avoid contention or you think about using a specialised data structure to avoid reimplementing your own. Note that it is both possible that @polytypic has encountered specific scenarios where a backoff in time was a solution, and at the same time that it is not desirable to have a backoff parameter for `Atomic.modify`.

As a concrete example of avoiding contention, instead of incrementing atomically the next object id, the runtime allocates object ids in chunks of 1024 per domain. (Trying, instead, to reduce cache-coherence traffic by giving way to other threads is pretty low on the list of good ideas.) As another illustration of contention-avoidance strategies, here you have [benchmarks for adaptive multi-counters](https://travisdowns.github.io/blog/2020/07/06/concurrency-costs.html#adaptive-multi-counter) (spreading a counter in space), which is pretty enlightening for its performance measurements. The best thing to do is probably very dependent on context and we have not yet seen evidence of backoff (in time) being a general solution (evidence which I have tried to find, independently of the various papers and blog posts @polytypic gave us which did not convince me).

Now trying to find a "default" backoff strategy sounds even harder given the lack of source material, but consider also that you would want it to be parametric in the user's scheduler (cf. similar discussions about a thread-safe Lazy).

As it stands though I have the impression that this whole discussion on backoff strategy stems from initially putting lock-free CAS loops and busy-waiting loops in the same bag. See e.g. the fact that the PR description mentions `Domainslib.Chan`, and `cpu_relax` as a backoff strategy (which is only advocated in Intel's optimization manual in relationship with spin locks). I am even more convinced than before that there should be no backoff parameter to avoid disseminating this confusion.

NoahTheDuke commented:
Coming from Clojure which has [atoms](https://clojure.org/reference/atoms) ("shared, synchronous, independent state"), `modify`/`modify_get` would be very helpful in writing thread-safe code.

A couple of thoughts on the discussion based on my experience with Atoms in Clojure:

1) I agree with @gadmm that there's no need for a backoff parameter. It will complicate the function implementation and confuse users as to the intended purpose of the function.

2) I think the function parameter order should be `[modify* r f]`. I know that they apply `f` to `r` so `[f r]` mirrors that, but 1) every other function in Atomics has the atomic first and keeping that consistent makes learning and usage easier, and 2) if you're writing an anonymous function, it will probably span multiple lines which obscures the targeted atomic. The `at_exit` change in this PR is an example of why the atomic should go first.

3) I think that `modify_get`'s signature should be `'a -> ('a -> 'a) -> 'a` which is to say, the return value of `f` is the new value of the atomic. First, returning a new value means that the "get" of `modify_get` has different semantics than the "get" of `get`, which returns the value of the atomic. That's confusing and inconsistent. Second (and more importantly), forcing `f` to return two objects, one of which will be returned to the caller, and the other is set as the new Atomic value, complicates all modify functions for the rare instance where you want a different return value than new state of the atomic. Requiring that `f` returns a pair means that you can't use most built-in functions anymore.

For example, if you wanted to atomically build a list of objects and check the length after, you couldn't write
```ocaml
let new_list = modify_get r (List.cons "Hello world") in
if 10 <= List.count new_list
  then func_1 new_list
  else func_2 new_list
```
You'd have to write
```ocaml
let new_list = modify_get r (fun l -> let new_list = List.cons "Hello world" l in new_list, new_list) in
if 10 <= List.count new_list
  then func_1 new_list
  else func_2 new_list
```
This is especially evident if you think of `fetch_and_add` as sugar over `modify_get`:
```ocaml
let fetch_and_add r i = modify_get r ((+) i)
```

Compare this with Clojure's `swap!` which returns the `new-val`:
```clojure
(let [new-list (swap! r cons "hello world")] ; prepend the string "hello world" to an atomic list
  (if (<= 10 (count new-list)) ; the length of the list is now 10 or more
    (func-1 new-list)
    (func-2 new-list)))
```

If you want a specialized return value that depends on the value of the old value while still modifying the new state (as in the docstring example), that could be a third function, something like `modify_pluck`.

gasche commented:
I decided to revisit this issue and write a microbenchmark to measure the impact of backoff strategies in `Atomic.modify`. 

I wrote a toy microbenchmark here: https://codeberg.org/gasche/bench-backoff/src/branch/trunk/bench_backoff.ml

The benchmark has a shared counter, and domains repeatedly increment it -- this is a case of extreme contention. A baseline uses `fetch_and_add` (should be the fastest), another uses a mutex and a non-atomic reference (should be the slowest), and then there is CAS without backoff, CAS with just `Domain.cpu_relax ()` as a backoff strategy and then CAS with exponential backoff as implemented by @polytypic in his [backoff](https://github.com/ocaml-multicore/backoff/tree/main) library.

Usage example:

```
make
hyperfine -L impl fetch_and_add,CAS_no_backoff,CAS_backoff_relax,CAS_backoff_exponential,mutex "NITERS=500_000 DOMAINS=2 IMPL={impl} ./bench_backoff.exe" --command-name {impl} --export-markdown dom2.md
```

Results with 2 domains (N_ITERS=5_000_000):

| Command (DOMAINS=2)       |    Mean [ms] |    Relative |
|:--------------------------|-------------:|------------:|
| `fetch_and_add`           |  106.7 ± 7.4 | 1.08 ± 0.08 |
| `CAS_no_backoff`          |  351.9 ± 4.8 | 3.55 ± 0.10 |
| `CAS_backoff_relax`       |  173.9 ± 9.9 | 1.75 ± 0.11 |
| `CAS_backoff_exponential` |   99.1 ± 2.4 |        1.00 |
| `mutex`                   | 678.6 ± 58.1 | 6.85 ± 0.61 |


Results with 12 domains (the number of cores on my machines) (N_ITERS=500_000):

| Command (DOMAINS=12)      |    Mean [ms] |    Relative |
|:--------------------------|-------------:|------------:|
| `fetch_and_add`           |  142.7 ± 6.0 |        1.00 |
| `CAS_no_backoff`          | 532.5 ± 27.1 | 3.73 ± 0.25 |
| `CAS_backoff_relax`       | 516.9 ± 20.2 | 3.62 ± 0.21 |
| `CAS_backoff_exponential` |  265.4 ± 9.2 | 1.86 ± 0.10 |
| `mutex`                   | 772.5 ± 14.3 | 5.41 ± 0.25 |

This microbenchmark tends to indicate that, even with a low number of domain, there is in fact a clear performance advantage to using an exponential backoff strategy under very high contention.

gasche committed:
restrict to just Atomic.modify for now

gasche committed:
Atomic: import the Backoff library as a submodule

gasche committed:
Atomic.Backoff: remove dependency on other stdlib modules

In particular, we stop doing *randomized* exponential backoff to
remove the dependency on the Random module, instead we use a
deterministic backoff of exactly 2^R iterations on round R.

gasche committed:
use Atomic.Backoff in the implementation of Atomic.modify

gasche commented:
I made the following changes to the PR:

- I removed `modify_get` whose prototype was not completely convincing, and include only `modify` for now. I would like to iterate to get agreement on that one function and its implementation, and then maybe consider reintroducing other functions later (after this is settled, or even in a separate PR.)

- I introduced exponential backoff following the encouraging results of my silly microbenchmark https://github.com/ocaml/ocaml/pull/10798#issuecomment-3590414254 . The implementation uses a vendored version of the [backoff](https://github.com/ocaml-multicore/backoff) library, modified to be deterministic rather than randomized to avoid dependency cycles within the stdlib. (Deterministic backoff improves the result on my microbenchmark, so I decided it was okay.)

gasche commented:
@NoahTheDuke thanks for your feedback. Comments inline.

> 1. I agree with @gadmm that there's no need for a backoff parameter. It will complicate the function implementation and confuse users as to the intended purpose of the function.

Ultimately I decided to add exponential backoff, but this is hidden in the implementation and does not complicate its usage.

> 2. I think the function parameter order should be `[modify* r f]`. I know that they apply `f` to `r` so `[f r]` mirrors that, but 1) every other function in Atomics has the atomic first and keeping that consistent makes learning and usage easier, and 2) if you're writing an anonymous function, it will probably span multiple lines which obscures the targeted atomic. The `at_exit` change in this PR is an example of why the atomic should go first.

I believe that the standard stdlib style is rather to have the function parameter first, for example `Foo.{map,iter}` takes the function first, even for mutable data structures, and this is arguably a form of `map_inplace` (which we have in Array). There is one exception, the `bind` function of monadic containers, because as you suggest putting the function last is more convenient, but I believe that adding more exceptions would be surprising.

> 3. I think that `modify_get`'s signature should be `'a -> ('a -> 'a) -> 'a` which is to say, the return value of `f` is the new value of the atomic. First, returning a new value means that the "get" of `modify_get` has different semantics than the "get" of `get`, which returns the value of the atomic. That's confusing and inconsistent. Second (and more importantly), forcing `f` to return two objects, one of which will be returned to the caller, and the other is set as the new Atomic value, complicates all modify functions for the rare instance where you want a different return value than new state of the atomic. Requiring that `f` returns a pair means that you can't use most built-in functions anymore.

Excellent points, thanks! My current best idea is to expose `Atomic.get_and_modify` and `Atomic.modify_and_get`. But for now I decided to keep things simple and just remove `modify_get`, focusing on `modify` for now.

redianthus commented:
@gasche, the [Hygiene / Checks (pull_request)](https://github.com/ocaml/ocaml/actions/runs/19789479716/job/56700502899?pr=10798) is unhappy. Also, a rebase is needed. (The PR's title could also be updated to remove `modify_get`).

redianthus commented on stdlib/atomic.mli:
Should probably be updated to use `Atomic.modify` along with the explanation right after `{2 Treiber Stack}`.

gasche replied:
This is a good remark but I am not sure whether we want to do it. Showing people how to implement lock-free algorithms using retry loops has merit as well. I may want to tweak the documentation to show both versions.

gasche replied:
I added this as an alternative way to write this example.

redianthus commented on stdlib/atomic.mli:
Ditto.

gasche replied:
I don't know if it is actually reasonable to implement `pop` using `Atomic.modify`. It would look something like this:

```ocaml
let pop t =
  let result = ref None in
  Atomic.modify (function
  | [] -> []
  | x::xs -> result := Some x; xs);
  !result
```

which is fairly meh. I would rather think of a more general API for that case, something like `('a -> 'a option * 'b)`.

gasche commented:
We discussed this PR at the last maintainer meeting (January 21st), and there was consensus in favor of the proposal, with just `modify` (for now?) and a reasonable (but not flexible) exponential-backoff strategy hidden in the implementation.

This still needs an implementation review and two approvals.

dbuenzli commented:
The standard library team had a look at this and is enthusiastic at having it in the toolbox. However rather than introduce new terminology (`modify`)  we propose to expose this under these names:

```ocaml
val Atomic.update : ('a -> 'a) -> 'a t -> unit 
val Atomic.map_update : ('a -> 'a * 'b) -> 'a t -> 'b
```

Which is rationalized as: 

1. `update` is already in [`Map.update`](https://ocaml.org/manual/5.3/api/Map.S.html#VALupdate) with similar functionality (functional update).
2. `map_update` follows this [function composition](https://github.com/ocaml/ocaml/pull/14185#issuecomment-3173775094) pattern often found in the `List` module. 

Some concerns were voiced that the original name and the proposed ones do not reflect the fact that the update function may be called more than once but nothing really suitable was found and we are happy with that.

gasche commented:
I am happy to replace `modify` with `update` following the advice.

For now I dropped `modify_get` from the PR so I would not re-include `map_update` yet. In the specific case of atomic operations I would note that there is already precedent with `fetch_and_add` and `add_and_fetch` for allowing people to observe both the pre-update and the post-update value as a return value; I think that it is actually useful to have both, and slightly uneasy with the fact that the use-cases of `modify_get` I found in the wild would duplicate computation to emulate the latter with the former, `Atomic.modify_get (fun x -> (x+1, x+1))`. I'm not sure your proposed naming scheme works well for that, and maybe we want to stay close to the different naming conventions of atomic operations across programming languages. But in any case, we don't need to decide this for now as `atomic_get` is not in scope anymore.

dbuenzli commented:
> with `fetch_and_add` and `add_and_fetch` for allowing people to observe both the pre-update and the post-update value as a return value

As far as I know only `Atomic.fetch_and_add` exists. Personally I find both these names not very enlightening they seem like `get` functions to me, no `set` functions. Note that `map_update` allows you to express both at the call site.

OlivierNicole commented:
It seems like there are no naming disagreements on the contents of the PR. If I understand correctly, the implementation hasn’t been reviewed yet. Do we have any reviewers for it?

If no one volunteers, I can add it to my to-review list, with no guarantees in terms of latency.

redianthus commented on stdlib/atomic.ml:
```diff
@@ -12,6 +13,101 @@
+       for-loop (uses two variables) to reduce register pressure.  Local ref does
+       not allocate with native compiler. *)
+    let t = ref (wait_mask land single_mask) in
+    while 0 <= !t do
```
Shouldn't it be `0 < !t` ? (It probably does not matter but I'm wondering if I overlooked something and there is actually a reason why it must be `<=`)

gasche replied:
`wait_mask` is of the form `2^R - 1`, so `t` is also of this form, and looping for `t + 1` iterations does give `2^R` iterations as documented. In the more general version of the Backoff module, where the number of iterations is chosen randomly, the extra iteration here guarantees that we wait between 1 and 2^R times, instead of waiting between 0 and 2^R-1 times.

gasche commented:
Note: I would be interested in eventually getting this merged, and for this we need implementation reviews. (There will be a "runtime meeting" next week so I can try to fish for reviewers there.)

(I had forgotten about the question of @redianthus, I'll have a look shortly.)

OlivierNicole commented:
I have started to review the implementation, to have a second review after @redianthus's (who posted several comments and not just one, by the way!).

Is it on purpose that the docstring of `Atomic.modify` does not document the exponential backoff and only says that `f` is called repeatedly?

gasche commented:
> Is it on purpose that the docstring of Atomic.modify does not document the exponential backoff and only says that `f` is called repeatedly?

This is not an intentional choice but I think that it is reasonable to leave exponential backoff as an implementation detail, instead of making it part of the specification. (It is non-obvious among people discussing the issue that it is a good choice; so it may change in the future.)

gasche commented:
(I rebased the library on top of the current trunk.)

dbuenzli commented:
It seems that [this comment](https://github.com/ocaml/ocaml/pull/10798#issuecomment-3853954400) was not taken into account.

gasche commented:
I had indeed forgotten about the suggestion to rename `modify` into `update`; apologies, and thanks for the reminder.

A nitpick note though: above I proposed using `'a option` to express the absence of update/modification, so potential extensions as `('a -> 'a option) -> 'a t -> unit` and `('a -> 'a option * 'b) -> 'a t -> 'b`. I wonder if this still works for `update`, where returning None is understood in the case of Map.update as "remove from the table" rather than "do not change things".

(Another option would be to forget about the `option` part and instead check whether the output is physically equal to the input, and not modify the reference in that case. But it is weirder and more implicit in use-sites.)

For example:

```ocaml
let pop stack =
  Atomic.update_map (function
  | [] -> None
  | x :: xs -> Some xs, x
  ) stack

let pop =
  Atomic.update_map (function
  | [] -> []
  | x::xs -> xs, x
  ) stack
```

dbuenzli commented:
I'm not sure what would work better in these cases with this new `modify` verb that you want to introduce here.

gasche committed:
rename Atomic.modify into Atomic.update

Suggested-by: Daniel Bünzli <daniel.buenzli@erratique.ch>

gasche committed:
Atomic.update: no-op on physical equality

gasche commented:
I renamed `Atomic.modify` into `Atomic.update`, and introduced a no-op path when the new value is physically equal to the previous one. (This is a valid refinement of the the non-deterministic set of behaviors allowed by the previous implementation.)

gasche commented:
I also considered re-introducing `Atomic.map_update`, but I decided that this was worth more discussion in a separate PR.

c-cube commented:
Please do reintroduce `map_update` :pray: (otherwise it risks landing in a different version of OCaml, which is never fun to bisect on the fated day of an opam release of a package depending on these new features)!

I'd still suggest the one [I use](https://github.com/c-cube/ocaml-containers/blob/c99883b841b52e78881ad14c2132dca823c0d6ec/src/core/CCAtomic.ml#L50) (note I don't even have the `unit`-returning version!) or at least its signature.

Adapted here it could look like this:

```ocaml

  let map_update f t =
    let rec loop ~backoff f t =
      let v_old = get t in
      let result, v_new = f v_old in
      if v_old == v_new ||  compare_and_set t v_old v_new
      then result
      else loop ~backoff:(Backoff.once backoff) f t
    in loop ~backoff:Backoff.default f t
```

gasche commented:
The problem of trying to do several things at once is that stuff never gets merged. (I submitted this PR in 2021!)

redianthus commented:
I agree, let's get this one first so it is not delayed even more.
Nothing prevents us to try getting `map_update` merged quickly in a separate PR.

OlivierNicole commented on stdlib/atomic.ml:
```diff
@@ -28,6 +124,15 @@ module Loc = struct
+    let rec loop ~backoff f t =
+      let v_old = get t in
+      let v_new = f v_old in
+      if v_old == v_new ||  compare_and_set t v_old v_new
```
Nit: two space characters

OlivierNicole commented on testsuite/tests/lib-atomic/test_atomic.ml:
```diff
@@ -37,3 +37,14 @@ let () =
+
+let () =
+  let r = Atomic.make 1 in
+  Atomic.modify (fun i ->
```
This should be updated to `Atomic.update`.

OlivierNicole reviewed (approved):
LGTM modulo fixing the tests (some renamings in Cmm in  `atomic-locs/record-fields`, and updating the function name in `lib-atomic/test_atomic`, see comment below).

gasche committed:
promote cmm stamps in the atomic-locs testsuite

gasche commented:
Thanks! I fixed the issues reported and rebased the PR. I count two approvals (from you and @redianthus as stdlib-committee representant) so this should be good to merge if the CI agrees.

dbuenzli commented:
Can you please update the PR title to reflect what is being merged (makes it easier to find stuff when it's time to go back in time).

gasche commented:
I learned today while reading the Rust 1.95.0 release notes that Rust stdlib has an `update` function for atomics that has almost exactly the same behavior as our new `Atomic.update` (and was just stabilized in 1.95.0):

https://doc.rust-lang.org/stable/std/sync/atomic/struct.AtomicPtr.html#method.update

```rust
    pub fn update(
        &self,
        set_order: Ordering,
        fetch_order: Ordering,
        mut f: impl FnMut(*mut T) -> *mut T,
    ) -> *mut T
```

beginning of the documentation:

> Fetches the value, applies a function to it that it return a new value.
> The new value is stored and the old value is returned.
> [...]
> Note: This may call the function multiple times if the value has been changed from other threads in
> the meantime, but the function will have been applied only once to the stored value.
